kernel_build.sh: Improve kernel build script

shreeya-patel98 opened the pull request:
Make some improvements to the kernel-build script for branch name related issues and also add an option to skip mrproper if needed.

PlaidCat commented on kernel_build.sh:
```diff
@@ -7,7 +7,7 @@ fi
 pwd
 
-BRANCH=$(git branch | grep \* | cut -d ' ' -f2 | sed -r 's/[{}/]/_/g')
+BRANCH=$(git branch | grep \* | cut -d ' ' -f2 | sed -r 's/[{}\/()]/_/g')
```
I've been doing something similar lately because with some of the `fips/sig-cloud` branch names + the SHA for the Local host name they're becoming longer than 64 characters and Make hates it.

This is what I've been adding but havn't made a PR about it yet

```
sed 's/[{}]//g; s/\//_/g'
```
note the `;` so sed can do multiple things i a single command

This way we just remove the brackets (can you'd your extension above because this ends up looking silly and wastefull when taking in to consideration the `LOCALVERSION` issue.

```
{jmaple)_fips-legacy-8/4.18*
_jmaple__fips-legacy-8_4.18*
```
My alteration above turns it into
`jmaple_fips-legacy-8_4.18*`

For context we started suing `{username}_<branch>` so lexicographical all the user branches are together.

shreeya-patel98 replied:
@PlaidCat would you like me to make these changes that you mentioned in addition to mine?

thefossguy-ciq replied:
Speaking for myself, I would like if you incorporated Maple's workaround in your PR. Would approve.

shreeya-patel98 replied:
@thefossguy-ciq This one is now outdated. The latest changes that I pushed also has Jonathan's workaround. Please check

thefossguy-ciq replied:
I didn't notice that. My bad, sorry.

thefossguy-ciq commented on kernel_build.sh:
```diff
@@ -12,12 +17,15 @@ BRANCH=$(git branch | grep \* | cut -d ' ' -f2 | sed -r 's/[{}\/()]/_/g')
 START=$(date +%s)
 START_MRPROPER=$(date +%s)
 if [ -e .config ]; then
```
You can also skip a `make mrproper` by removing the `.config` file. That's what i do.

PlaidCat replied:
Requires knowing the underlying script logic, bu t a fair point

shreeya-patel98 replied:
But that's not the ideal way to do make mrproper right? You might want to keep the config file as it is and skip mrproper to avoid building the whole kernel again. I usually don't do mrproper if it's not needed, it saves a lot of build time since most of the modules and object files will be same as before ( only the ones we make changes to needs to be recompiled )

PlaidCat commented on kernel_build.sh:
```diff
@@ -3,6 +3,11 @@
+    shift
+fi
+
+if [ "$1" == "skip-mrproper" ]; then
+    SKIP_MRPROPER=true
```
I think this is fine for now, but there will come moments when we do actually need to skip `mrproper` and `skipkabi`. this is mostly related to the GHK long term kernels.

This comment is non-blocking. and maybe a clean up later.

We should probably start looking at introducing `getopts`
https://zerotomastery.io/blog/bash-getopts/

thefossguy-ciq replied:
I use environment variables to not deal with argument parsing in bash but I agree that `getopts` is the way to go here.

shreeya-patel98 replied:
yeah I agree, let me make this change to the script to start using getopts

shreeya-patel98 committed:
kernel_build.sh: Improve branch name sanitization

Extended sed pattern to replace some more characters
which could be in a branch name

Signed-off-by: Shreeya Patel <spatel@ciq.com>

shreeya-patel98 committed:
kernel_build.sh: Use getopts for argument parsing

Use getopts for argument parsing and add options
to skip kabi and mrproper whenever needed.

Signed-off-by: Shreeya Patel <spatel@ciq.com>

thefossguy-ciq commented on kernel_build.sh:
```diff
@@ -1,31 +1,63 @@
 START=$(date +%s)
 START_MRPROPER=$(date +%s)
+
 if [ -e .config ]; then
```
I wonder if it is appropriate to skip this check to see if a `.config` file exists or not. That has no impact on the result of running `make mrproper` and shouldn't have an impact on the decision of weather to run `make mrproper` down the line (because the check for `$SKIP_MRPROPER` is inside this conditional).

```
[pratham@lts-8-6 ppatel_ciqlts8_6]$ file .config
.config: cannot open `.config' (No such file or directory)
[pratham@lts-8-6 ppatel_ciqlts8_6]$ make mrproper
[pratham@lts-8-6 ppatel_ciqlts8_6]$ echo $?
0
[pratham@lts-8-6 ppatel_ciqlts8_6]$
```

**The rest looks good!**

PlaidCat replied:
I would concur with this we can just get rid of this if else conditional

thefossguy-ciq commented on kernel_build.sh:
```diff
@@ -1,31 +1,59 @@
+    if [[ $? -ne 0 ]]; then
+	echo "Error: make mrproper failed"
+	echo "[TIMER]{MRPROPER} $(( $(date +%s) - $START_MRPROPER ))s"
+	exit 1
```
Final nit: Could you please replace these tabs with spaces?

thefossguy-ciq reviewed (approved):
Thank you for all changes and sorry about my OCD.
🚤

PlaidCat reviewed (approved):
:shipit:
